### PR TITLE
[CB] Test continuous batching through the scheduler

### LINKED ISSUE #167: [CB] Test corner cases
* scenario where the tkv comes to the end of 2048 block, a new batch should start
* two sequences finish at the same time
* new request arrives just when another finishes, it should be scheduled directly

### PR DESCRIPTION
**Pull request by sducouedic**
- Closes #167 , #166, #165

### CODE REVIEW

File: tests/e2e/test_spyre_cb.py [hunk 1]
```diff
@@ -324,7 +327,298 @@ def get_params_test_blocks_borders_misaligned_prompts():
...
+            "step": 58,
+            "tkv": 184,
+            "waiting": ["2"],
+            "running": ["0"],
+            "request_outputs": ["1", "0"],
+            "finished_requests": ["1"]
+        },
+        {
+            # Decode sequence 0
+            # Cannot decode sequence 2: 185 + 80 = 265 > 256
```
Comment (prashantgupta24) [inline tests/e2e/test_spyre_cb.py:567]: should this say `prefill` instead?
```suggestion
            # Cannot prefill sequence 2: 185 + 80 = 265 > 256
```

**Comment by prashantgupta24** [comment]
I like it!

File: tests/e2e/test_spyre_cb.py (hunk 1 above)
Comment (sducouedic) [reply tests/e2e/test_spyre_cb.py]: good catch!

**Review by prashantgupta24 — APPROVED** [review]
LGTM. I guess the left padding question still remains.

File: tests/e2e/test_spyre_cb.py [hunk 2]
```diff
@@ -618,7 +623,129 @@ def get_params_test_scheduler_constraints_max_prompt_len():
...
+            "waiting": ["2"],
+            "running": ["1"],
+            "request_outputs": ["1", "0"],
+            "finished_requests": ["0"]
+        },
+        {
+            # Prefill sequence 2
+            # TODO @Yannick, should the left padding ideally already be removed?
+            "step": 42,
+            "tkv": 103,  # <-- should be 64?
```
[the remark this answers is not in the document]
Comment (sducouedic) [reply tests/e2e/test_spyre_cb.py:700]: you're other comment is correct, it is about calling `reduce_left_padding`

File: tests/e2e/test_spyre_cb.py [hunk 3]
```diff
@@ -618,7 +623,129 @@ def get_params_test_scheduler_constraints_max_prompt_len():
...
+            "step": 41,
+            "tkv": 103,
+            "waiting": ["2"],
+            "running": ["1"],
+            "request_outputs": ["1", "0"],
+            "finished_requests": ["0"]
+        },
+        {
+            # Prefill sequence 2
+            # TODO @Yannick, should the left padding ideally already be removed?
```
[the remark this answers is not in the document]
Comment (sducouedic) [reply tests/e2e/test_spyre_cb.py:698]: Yes, you're right! I am wondering if it is the desired behaviour, and if something prevents to have the left padding already removed in the subsequent prefill. Conceptually it makes more sense to have the padding already removed if a prefill happens (ie. call `reduce_left_padding` in both decode and prefill). Also supposing the tkv is too close to `max_model_len` at the time of the prefill, the prompt would need to wait 1 decode, instead of being directly scheduled if the padding reduction happens directly.

[the remark this answers is not in the document]
Comment (sducouedic) [reply tests/e2e/test_spyre_cb.py:698]: ah I see what you mean, that would be more code instead of just calling the `reduce_left_padding` function after adding the padding. Let's see what @yannicks1 has to say

File: tests/e2e/test_spyre_cb.py [hunk 4]
```diff
@@ -324,7 +329,423 @@ def get_params_test_blocks_borders_misaligned_prompts():
...
+    checked_steps = [
+        {
+            "step": 0,
+            "tkv": 0,
+            "waiting": ["0", "1"],
+            "running": [],
+            "request_outputs": []
+        },
+        {
+            "step": 1,
```
Comment (yannicks1) [inline tests/e2e/test_spyre_cb.py:446]:
```suggestion
            # Prefill sequence 0
            "step": 1,
```

Comment (yannicks1) [reply tests/e2e/test_spyre_cb.py:446]: maybe add this comment here for completeness.

File: tests/e2e/test_spyre_cb.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yannicks1) [reply tests/e2e/test_spyre_cb.py:698]: I do agree and am working on reducing the left padding in every scheduler step (prefil and decode). Thanks for pointing this out!

**Review by yannicks1 — APPROVED** [review]
This is really nice work @sducouedic. Super clean tests on scheduler step level. I suggest to merge this. I am already preparing the improved reducing of the left padding on another branch and would then update the tests accordingly.
